crates/kornia-3d/src/voxel_grid.rs

```diff
@@ -0,0 +1,382 @@
+use crate::pointcloud::PointCloud;
+use crate::vector::DVec3;
+use std::collections::HashMap;
+
+/// Type alias for voxel data in the grid when downsampling all data.
+type VoxelData = (DVec3, Vec<[u8; 3]>, Vec<[f64; 3]>, usize);
+
+/// A 3D voxel grid for downsampling point clouds.
+#[derive(Debug, Clone)]
+pub struct VoxelGrid {
+    /// The size of the voxel (leaf) in x, y, z dimensions.
+    leaf_size: DVec3,
+    /// Minimum number of points required per voxel.
+    min_points_per_voxel: usize,
+    /// Whether to downsample all data (points, colors, normals) or just points.
+    downsample_all_data: bool,
+}
+
+impl VoxelGrid {
+    /// Creates a new `VoxelGrid` with the specified leaf size.
+    ///
+    /// # Arguments
+    /// * `leaf_size` - The size of the voxel in x, y, z dimensions.
+    ///
+    /// # Panics
+    /// Panics if any dimension of `leaf_size` is non-positive.
+    pub fn new(leaf_size: DVec3) -> Self {
+        if leaf_size.x <= 0.0 || leaf_size.y <= 0.0 || leaf_size.z <= 0.0 {
+            panic!("Leaf size must be positive in all dimensions");
+        }
+
+        VoxelGrid {
+            leaf_size,
+            min_points_per_voxel: 1,
+            downsample_all_data: true,
+        }
+    }
+
+    /// Downsamples the input point cloud by grouping points into voxels and computing centroids.
+    ///
+    /// # Arguments
+    /// * `point_cloud` - The input point cloud to downsample.
+    ///
+    /// # Returns
+    /// A new `PointCloud` containing the downsampled points, and optionally colors and normals.
+    pub fn downsample(&mut self, point_cloud: &PointCloud) -> PointCloud {
+        let mut grid: HashMap<(i32, i32, i32), VoxelData> = HashMap::new();
+
+        // Group points into voxels
+        for (i, point) in point_cloud.points().iter().enumerate() {
+            let point_vec = DVec3::from_array(point);
+            let ix = (point_vec.x / self.leaf_size.x).floor() as i32;
+            let iy = (point_vec.y / self.leaf_size.y).floor() as i32;
+            let iz = (point_vec.z / self.leaf_size.z).floor() as i32;
+            let key = (ix, iy, iz);
+            let entry = grid.entry(key).or_insert((
+                DVec3::from_array(&[0.0, 0.0, 0.0]),
+                Vec::new(),
+                Vec::new(),
+                0,
+            ));
+            entry.0.x += point_vec.x;
+            entry.0.y += point_vec.y;
+            entry.0.z += point_vec.z;
+            entry.3 += 1;
+
+            if self.downsample_all_data {
```

Comment:
something like this would be more compact and rust idiomatic

```rust
if !self.downsample_all_data {
    return;
}
if let Some(&color)  = point_cloud.colors().and_then(|c| c.get(i)) {
    entry.1.push(color);
}
if let Some(&normal) = point_cloud.normals().and_then(|n| n.get(i)) {
    entry.2.push(normal);
}
```